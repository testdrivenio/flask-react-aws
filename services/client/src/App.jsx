import React, { Component } from "react";
import axios from "axios";
import { Route, Switch } from "react-router-dom";
import Modal from "react-modal";

import UsersList from "./components/UsersList";
import About from "./components/About";
import NavBar from "./components/NavBar";
import LoginForm from "./components/LoginForm";
import RegisterForm from "./components/RegisterForm";
import UserStatus from "./components/UserStatus";
import Message from "./components/Message";
import AddUser from "./components/AddUser";

// new
const modalStyles = {
  content: {
    top: "0",
    left: "0",
    right: "0",
    bottom: "0",
    border: 0,
    background: "transparent"
  }
};

Modal.setAppElement(document.getElementById("root"));

class App extends Component {
  constructor() {
    super();

    this.state = {
      users: [],
      title: "TestDriven.io",
      accessToken: null,
      messageType: null,
      messageText: null,
      showModal: false
    };
  }

  componentDidMount() {
    this.getUsers();
  }

  getUsers = () => {
    axios
      .get(`${process.env.REACT_APP_USERS_SERVICE_URL}/users`)
      .then(res => {
        this.setState({ users: res.data });
      })
      .catch(err => {
        console.log(err);
      });
  };

  addUser = data => {
    axios
      .post(`${process.env.REACT_APP_USERS_SERVICE_URL}/users`, data)
      .then(res => {
        this.getUsers();
        this.setState({ username: "", email: "" });
        this.handleCloseModal();
        this.createMessage("success", "User added.");
      })
      .catch(err => {
        console.log(err);
        this.handleCloseModal();
        this.createMessage("danger", "That user already exists.");
      });
  };

  removeUser = user_id => {
    axios
      .delete(`${process.env.REACT_APP_USERS_SERVICE_URL}/users/${user_id}`)
      .then(res => {
        this.getUsers();
        this.createMessage("success", "User removed.");
      })
      .catch(err => {
        console.log(err);
        this.createMessage("danger", "Something went wrong.");
      });
  };

  handleRegisterFormSubmit = data => {
    const url = `${process.env.REACT_APP_USERS_SERVICE_URL}/auth/register`;
    axios
      .post(url, data)
      .then(res => {
        console.log(res.data);
        this.createMessage("success", "You have registered successfully.");
      })
      .catch(err => {
        console.log(err);
        this.createMessage("danger", "That user already exists.");
      });
  };

  handleLoginFormSubmit = data => {
    const url = `${process.env.REACT_APP_USERS_SERVICE_URL}/auth/login`;
    axios
      .post(url, data)
      .then(res => {
        this.setState({ accessToken: res.data.access_token });
        this.getUsers();
        window.localStorage.setItem("refreshToken", res.data.refresh_token);
        this.createMessage("success", "You have logged in successfully.");
      })
      .catch(err => {
        console.log(err);
        this.createMessage("danger", "Incorrect email and/or password.");
      });
  };

  isAuthenticated = () => {
    if (this.state.accessToken || this.validRefresh()) {
      return true;
    }
    return false;
  };

  validRefresh = () => {
    const token = window.localStorage.getItem("refreshToken");
    if (token) {
      axios
        .post(`${process.env.REACT_APP_USERS_SERVICE_URL}/auth/refresh`, {
          refresh_token: token
        })
        .then(res => {
          this.setState({ accessToken: res.data.access_token });
          this.getUsers();
          window.localStorage.setItem("refreshToken", res.data.refresh_token);
          return true;
        })
        .catch(err => {
          return false;
        });
    }
    return false;
  };

  logoutUser = () => {
    window.localStorage.removeItem("refreshToken");
    this.setState({ accessToken: null });
    this.createMessage("success", "You have logged out.");
  };

  createMessage = (type, text) => {
    this.setState({
      messageType: type,
      messageText: text
    });
    setTimeout(() => {
      this.removeMessage();
    }, 3000);
  };

  removeMessage = () => {
    this.setState({
      messageType: null,
      messageText: null
    });
  };

  handleOpenModal = () => {
    this.setState({ showModal: true });
  };

  handleCloseModal = () => {
    this.setState({ showModal: false });
  };

  render() {
    return (
      <div>
        <NavBar
          title={this.state.title}
          logoutUser={this.logoutUser}
          isAuthenticated={this.isAuthenticated}
        />
        <section className="section">
          <div className="container">
            {this.state.messageType && this.state.messageText && (
              <Message
                messageType={this.state.messageType}
                messageText={this.state.messageText}
                removeMessage={this.removeMessage}
              />
            )}
            <div className="columns">
              <div className="column is-half">
                <br />
                <Switch>
                  <Route
                    exact
                    path="/"
                    render={() => (
                      <div>
                        <h1 className="title is-1">Users</h1>
                        <hr />
                        <br />
                        {this.isAuthenticated() && (
                          <button
                            onClick={this.handleOpenModal}
                            className="button is-primary"
                          >
                            Add User
                          </button>
                        )}
                        <br />
                        <br />
                        <Modal
                          isOpen={this.state.showModal}
                          style={modalStyles}
                        >
                          <div className="modal is-active">
                            <div className="modal-background" />
                            <div className="modal-card">
                              <header className="modal-card-head">
                                <p className="modal-card-title">Add User</p>
                                <button
                                  className="delete"
                                  aria-label="close"
                                  onClick={this.handleCloseModal}
                                />
                              </header>
                              <section className="modal-card-body">
                                <AddUser addUser={this.addUser} />
                              </section>
                            </div>
                          </div>
                        </Modal>
                        <UsersList
                          users={this.state.users}
                          removeUser={this.removeUser}
                          isAuthenticated={this.isAuthenticated}
                        />
                      </div>
                    )}
                  />
                  <Route exact path="/about" component={About} />
                  <Route
                    exact
                    path="/register"
                    render={() => (
                      <RegisterForm
                        // eslint-disable-next-line react/jsx-handler-names
                        handleRegisterFormSubmit={this.handleRegisterFormSubmit}
                        isAuthenticated={this.isAuthenticated}
                      />
                    )}
                  />
                  <Route
                    exact
                    path="/login"
                    render={() => (
                      <LoginForm
                        // eslint-disable-next-line react/jsx-handler-names
                        handleLoginFormSubmit={this.handleLoginFormSubmit}
                        isAuthenticated={this.isAuthenticated}
                      />
                    )}
                  />
                  <Route
                    exact
                    path="/status"
                    render={() => (
                      <UserStatus
                        accessToken={this.state.accessToken}
                        isAuthenticated={this.isAuthenticated}
                      />
                    )}
                  />
                </Switch>
              </div>
            </div>
          </div>
        </section>
      </div>
    );
  }
}

export default App;
