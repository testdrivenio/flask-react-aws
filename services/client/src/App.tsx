import React, { useState, useEffect } from "react";
import { ChakraProvider, useDisclosure } from "@chakra-ui/react";
import { Route, Routes } from "react-router-dom";
import axios from "axios";
import Users from "./components/Users";
import About from "./components/About";
import NavBar from "./components/NavBar";
import LoginForm from "./components/LoginForm";
import RegisterForm from "./components/RegisterForm";
import UserStatus from "./components/UserStatus";
import Message from "./components/Message";
import AddUserModal from "./components/AddUserModal";

interface User {
  created_date: string;
  email: string;
  id: number;
  username: string;
}

const App = () => {
  const [users, setUsers] = useState<User[]>([]);
  const [title] = useState("TestDriven.io");
  const [accessToken, setAccessToken] = useState<string | null>(null);
  const [messageType, setMessageType] = useState<
    "info" | "warning" | "success" | "error" | null
  >(null);
  const [messageText, setMessageText] = useState<string | null>(null);
  const { isOpen, onOpen, onClose } = useDisclosure();

  useEffect(() => {
    const checkAuth = async () => {
      await validRefresh();
    };
    checkAuth();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const isAuthenticated = () => {
    return !!accessToken;
  };

  const fetchUsers = async () => {
    try {
      const response = await axios.get(
        `${import.meta.env.VITE_API_SERVICE_URL}/users`,
      );
      if (response.status === 200) {
        setUsers(response.data);
      }
    } catch (error) {
      console.error("Error fetching users:", error);
    }
  };

  const handleRegisterFormSubmit = async (data: {
    username: string;
    email: string;
    password: string;
  }) => {
    try {
      const url = `${import.meta.env.VITE_API_SERVICE_URL}/auth/register`;
      const response = await axios.post(url, data);
      console.log(response.data);
      createMessage("success", "Registration successful! You can now log in."); // Display success message
    } catch (err) {
      console.log(err);
      createMessage(
        "error",
        "Registration failed. The user might already exist.",
      ); // Display error message
    }
  };

  const handleLoginFormSubmit = async (data: {
    email: string;
    password: string;
  }) => {
    try {
      const url = `${import.meta.env.VITE_API_SERVICE_URL}/auth/login`;
      const response = await axios.post(url, data);
      console.log(response.data);
      setAccessToken(response.data.access_token);
      window.localStorage.setItem("refreshToken", response.data.refresh_token);
      await fetchUsers();
      createMessage("success", "Login successful!"); // Display success message
    } catch (err) {
      console.log(err);
      createMessage("error", "Login failed. Please check your credentials."); // Display error message
    }
  };

  const validRefresh = async () => {
    const token = window.localStorage.getItem("refreshToken");
    if (token) {
      try {
        const response = await axios.post(
          `${import.meta.env.VITE_API_SERVICE_URL}/auth/refresh`,
          {
            refresh_token: token,
          },
        );
        setAccessToken(response.data.access_token);
        await fetchUsers();
        window.localStorage.setItem(
          "refreshToken",
          response.data.refresh_token,
        );
        return true;
      } catch (err) {
        console.log(err);
        return false;
      }
    }
    return false;
  };

  const logoutUser = () => {
    setAccessToken(null);
    window.localStorage.removeItem("refreshToken");
    createMessage("info", "You have been logged out."); // Display success message
  };

  const clearMessage = () => {
    setMessageType(null);
    setMessageText(null);
  };

  const createMessage = (
    type: "info" | "warning" | "success" | "error",
    text: string,
  ) => {
    setMessageType(type);
    setMessageText(text);
    setTimeout(() => {
      clearMessage();
    }, 3000);
  };

  // Add a test message when the component mounts
  useEffect(() => {
    createMessage("info", "Welcome to the application!");
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const addUser = async (userData: {
    username: string;
    email: string;
    password: string;
  }) => {
    try {
      const response = await axios.post(
        `${import.meta.env.VITE_API_SERVICE_URL}/users`,
        {
          username: userData.username,
          email: userData.email,
          password: userData.password,
          created_date: new Date().toISOString(),
        },
      );

      // Update the users state with the new user
      setUsers((prevUsers) => [...prevUsers, response.data]);
      createMessage("success", "User added successfully.");
      onClose(); // Close the modal
      await fetchUsers(); // Fetch the updated list of users
    } catch (err) {
      console.error(err);
      createMessage(
        "error",
        "Failed to add user. The user might already exist.",
      );
    }
  };

  const removeUser = async (userId: number) => {
    try {
      await axios.delete(
        `${import.meta.env.VITE_API_SERVICE_URL}/users/${userId}`,
      );
      await fetchUsers(); // Fetch the updated list of users
      createMessage("success", "User removed successfully.");
    } catch (err) {
      console.error(err);
      createMessage("error", "Failed to remove user. Please try again.");
    }
  };

  return (
    <ChakraProvider>
      <NavBar
        title={title}
        logoutUser={logoutUser}
        isAuthenticated={isAuthenticated}
      />
      {messageType && messageText && (
        <Message
          messageType={messageType}
          messageText={messageText}
          onClose={clearMessage} // Make sure this is passed
        />
      )}
      <Routes>
        <Route
          path="/"
          element={
            <>
              <Users
                users={users}
                onAddUser={onOpen}
                removeUser={removeUser}
                isAuthenticated={isAuthenticated()}
              />
              <AddUserModal
                isOpen={isOpen}
                onClose={onClose}
                addUser={addUser}
              />
            </>
          }
        />
        <Route path="/about" element={<About />} />
        <Route
          path="/register"
          element={
            <RegisterForm
              onSubmit={handleRegisterFormSubmit}
              isAuthenticated={isAuthenticated}
            />
          }
        />
        <Route
          path="/login"
          element={
            <LoginForm
              onSubmit={handleLoginFormSubmit}
              isAuthenticated={isAuthenticated}
            />
          }
        />
        <Route
          path="/status"
          element={
            <UserStatus
              accessToken={accessToken || ""}
              isAuthenticated={isAuthenticated}
            />
          }
        />
      </Routes>
    </ChakraProvider>
  );
};

export default App;
