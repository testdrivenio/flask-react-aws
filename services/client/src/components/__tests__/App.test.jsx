import React from 'react';
import { cleanup } from '@testing-library/react';

import App from '../../App';

afterEach(cleanup);

it("renders", () => {
  const { asFragment } = renderWithRouter(<App />);
  expect(asFragment()).toMatchSnapshot();
});
