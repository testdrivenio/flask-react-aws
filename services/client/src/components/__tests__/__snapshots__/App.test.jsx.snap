// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <div>
    <nav
      aria-label="main navigation"
      class="navbar is-dark"
      role="navigation"
    >
      <section
        class="container"
      >
        <div
          class="navbar-brand"
        >
          <a
            class="navbar-item nav-title"
            href="/"
            style="font-weight: bold;"
          >
            TestDriven.io
          </a>
          <span
            class="nav-toggle navbar-burger"
          >
            <span />
            <span />
            <span />
          </span>
        </div>
        <div
          class="navbar-menu"
        >
          <div
            class="navbar-start"
          >
            <a
              class="navbar-item"
              data-testid="nav-about"
              href="/about"
            >
              About
            </a>
          </div>
          <div
            class="navbar-end"
          >
            <a
              class="navbar-item"
              data-testid="nav-register"
              href="/register"
            >
              Register
            </a>
            <a
              class="navbar-item"
              data-testid="nav-login"
              href="/login"
            >
              Log In
            </a>
          </div>
        </div>
      </section>
    </nav>
    <section
      class="section"
    >
      <div
        class="container"
      >
        <div
          class="columns"
        >
          <div
            class="column is-half"
          >
            <br />
            <div>
              <h1
                class="title is-1"
              >
                Users
              </h1>
              <hr />
              <br />
              <br />
              <br />
              <div>
                <table
                  class="table is-hoverable is-fullwidth"
                >
                  <thead>
                    <tr>
                      <th>
                        ID
                      </th>
                      <th>
                        Email
                      </th>
                      <th>
                        Username
                      </th>
                    </tr>
                  </thead>
                  <tbody />
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;
